File: app/src/main/java/com/bongorian/signa1/LivePerformance.java
package com.bongorian.signa1;

/** Continuous fault-time evolution and modulation. Camera/encoder clocks remain untouched. */
final class LivePerformance {
    static final int NATURAL=0,PULSE=1,SWELL=2,BURST=3,CASCADE=4;
    static final int FREE=0,LOOP=1,PING_PONG=2,STEP=3;
    final int style,clock;final float speed,tempo,depth,beats,division,width,chance,periodSeconds,stepSeconds;final boolean hold;
    LivePerformance(int style,int clock,float speed,float tempo,float depth,float beats,float division,float width,float chance,boolean hold){
        this(style,clock,speed,tempo,depth,beats,division,width,chance,hold,60f/tempo*beats,60f/tempo/division);
    }
    LivePerformance(int style,int clock,float speed,float tempo,float depth,float beats,float division,float width,float chance,boolean hold,float periodSeconds,float stepSeconds){
        this.periodSeconds=FaultModel.clamp(periodSeconds,.25f,32);this.stepSeconds=FaultModel.clamp(stepSeconds,.015625f,2);
        this.style=Math.max(0,Math.min(4,style));this.clock=Math.max(0,Math.min(3,clock));this.speed=FaultModel.clamp(speed,-4,4);this.tempo=FaultModel.clamp(tempo,30,240);this.depth=EffectParameters.unit(depth);this.beats=FaultModel.clamp(beats,1,16);this.division=FaultModel.clamp(division,1,16);this.width=FaultModel.clamp(width,.05f,.95f);this.chance=EffectParameters.unit(chance);this.hold=hold;
    }
    static LivePerformance defaults(){return new LivePerformance(NATURAL,FREE,1,90,.75f,4,4,.3f,.65f,false);}
    LivePerformance held(boolean value){return new LivePerformance(style,clock,speed,tempo,depth,beats,division,width,chance,value,periodSeconds,stepSeconds);}
    LivePerformance with(String key,float value){
        switch(key){
            case "period":return new LivePerformance(style,clock,speed,tempo,depth,beats,division,width,chance,hold,value,stepSeconds);
            case "interval":return new LivePerformance(style,clock,speed,tempo,depth,beats,division,width,chance,hold,periodSeconds,value);
            case "style":return new LivePerformance(Math.round(value),clock,speed,tempo,depth,beats,division,width,chance,hold,periodSeconds,stepSeconds);
            case "clock":return new LivePerformance(style,Math.round(value),speed,tempo,depth,beats,division,width,chance,hold,periodSeconds,stepSeconds);
            case "speed":return new LivePerformance(style,clock,value,tempo,depth,beats,division,width,chance,hold,periodSeconds,stepSeconds);
            case "tempo":return new LivePerformance(style,clock,speed,value,depth,beats,division,width,chance,hold,60f/value*beats,60f/value/division);
            case "depth":return new LivePerformance(style,clock,speed,tempo,value,beats,division,width,chance,hold,periodSeconds,stepSeconds);
            case "beats":return new LivePerformance(style,clock,speed,tempo,depth,value,division,width,chance,hold,60f/tempo*value,stepSeconds);
            case "division":return new LivePerformance(style,clock,speed,tempo,depth,beats,value,width,chance,hold,periodSeconds,60f/tempo/value);
            case "width":return new LivePerformance(style,clock,speed,tempo,depth,beats,division,value,chance,hold,periodSeconds,stepSeconds);
            case "chance":return new LivePerformance(style,clock,speed,tempo,depth,beats,division,width,value,hold,periodSeconds,stepSeconds);
            default:throw new IllegalArgumentException("Performance control");
        }
    }
    static double wrap(double value,double span){return value-Math.floor(value/span)*span;}
    double time(double position){double length=periodSeconds;switch(clock){case LOOP:return wrap(position,length);case PING_PONG:double phase=wrap(position,length*2);return phase<=length?phase:length*2-phase;case STEP:double step=stepSeconds;return Math.floor(position/step)*step;default:return position;}}
    float envelope(double seconds,int rank,int count,long seed){
        double cycle=seconds/periodSeconds,phase=wrap(cycle,1);float envelope;
        switch(style){
            case PULSE:envelope=(float)(.5-.5*Math.cos(phase*Math.PI*2));break;
            case SWELL:envelope=(float)phase;break;
            case BURST:
                if(FaultModel.random(seed^FaultModel.mix((long)Math.floor(cycle)))>=chance||phase>=width)envelope=0;
                else envelope=FaultModel.clamp((float)Math.min(phase/.025,(width-phase)/.075),0,1);break;
            case CASCADE:
                double distance=Math.abs(phase-rank/(double)Math.max(1,count));distance=Math.min(distance,1-distance);envelope=(float)Math.max(0,1-distance*Math.max(1,count));break;
            default:return 1;
        }
        return 1-depth+depth*envelope;
    }
    boolean warped(){return clock!=FREE||speed!=1||hold;}
}

File: app/src/main/kotlin/com/bongorian/signa1/LivePerformance.kt
package com.bongorian.signa1

import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

/** Continuous fault-time evolution and modulation. Camera/encoder clocks remain untouched. */
internal class LivePerformance
constructor(
    style: Int,
    clock: Int,
    speed: Float,
    tempo: Float,
    depth: Float,
    beats: Float,
    division: Float,
    width: Float,
    chance: Float,
    val hold: Boolean,
    periodSeconds: Float = 60f / tempo * beats,
    stepSeconds: Float = 60f / tempo / division,
) {
    val style: Int
    val clock: Int
    val speed: Float
    val tempo: Float
    val depth: Float
    val beats: Float
    val division: Float
    val width: Float
    val chance: Float
    val periodSeconds: Float
    val stepSeconds: Float

    init {
        this.periodSeconds = FaultModel.clamp(periodSeconds, .25f, 32f)
        this.stepSeconds = FaultModel.clamp(stepSeconds, .015625f, 2f)
        this.style = max(0, min(4, style))
        this.clock = max(0, min(3, clock))
        this.speed = FaultModel.clamp(speed, -4f, 4f)
        this.tempo = FaultModel.clamp(tempo, 30f, 240f)
        this.depth = EffectParameters.unit(depth)
        this.beats = FaultModel.clamp(beats, 1f, 16f)
        this.division = FaultModel.clamp(division, 1f, 16f)
        this.width = FaultModel.clamp(width, .05f, .95f)
        this.chance = EffectParameters.unit(chance)
    }

    fun held(value: Boolean): LivePerformance {
        return LivePerformance(
            style,
            clock,
            speed,
            tempo,
            depth,
            beats,
            division,
            width,
            chance,
            value,
            periodSeconds,
            stepSeconds,
        )
    }

    fun with(key: String, value: Float): LivePerformance {
        when (key) {
            "period" ->
                return LivePerformance(
                    style,
                    clock,
                    speed,
                    tempo,
                    depth,
                    beats,
                    division,
                    width,
                    chance,
                    hold,
                    value,
                    stepSeconds,
                )

            "interval" ->
                return LivePerformance(
                    style,
                    clock,
                    speed,
                    tempo,
                    depth,
                    beats,
                    division,
                    width,
                    chance,
                    hold,
                    periodSeconds,
                    value,
                )

            "style" ->
                return LivePerformance(
                    Math.round(value),
                    clock,
                    speed,
                    tempo,
                    depth,
                    beats,
                    division,
                    width,
                    chance,
                    hold,
                    periodSeconds,
                    stepSeconds,
                )

            "clock" ->
                return LivePerformance(
                    style,
                    Math.round(value),
                    speed,
                    tempo,
                    depth,
                    beats,
                    division,
                    width,
                    chance,
                    hold,
                    periodSeconds,
                    stepSeconds,
                )

            "speed" ->
                return LivePerformance(
                    style,
                    clock,
                    value,
                    tempo,
                    depth,
                    beats,
                    division,
                    width,
                    chance,
                    hold,
                    periodSeconds,
                    stepSeconds,
                )

            "tempo" ->
                return LivePerformance(
                    style,
                    clock,
                    speed,
                    value,
                    depth,
                    beats,
                    division,
                    width,
                    chance,
                    hold,
                    60f / value * beats,
                    60f / value / division,
                )

            "depth" ->
                return LivePerformance(
                    style,
                    clock,
                    speed,
                    tempo,
                    value,
                    beats,
                    division,
                    width,
                    chance,
                    hold,
                    periodSeconds,
                    stepSeconds,
                )

            "beats" ->
                return LivePerformance(
                    style,
                    clock,
                    speed,
                    tempo,
                    depth,
                    value,
                    division,
                    width,
                    chance,
                    hold,
                    60f / tempo * value,
                    stepSeconds,
                )

            "division" ->
                return LivePerformance(
                    style,
                    clock,
                    speed,
                    tempo,
                    depth,
                    beats,
                    value,
                    width,
                    chance,
                    hold,
                    periodSeconds,
                    60f / tempo / value,
                )

            "width" ->
                return LivePerformance(
                    style,
                    clock,
                    speed,
                    tempo,
                    depth,
                    beats,
                    division,
                    value,
                    chance,
                    hold,
                    periodSeconds,
                    stepSeconds,
                )

            "chance" ->
                return LivePerformance(
                    style,
                    clock,
                    speed,
                    tempo,
                    depth,
                    beats,
                    division,
                    width,
                    value,
                    hold,
                    periodSeconds,
                    stepSeconds,
                )

            else -> throw IllegalArgumentException("Performance control")
        }
    }

    fun time(position: Double): Double {
        val length = periodSeconds.toDouble()
        when (clock) {
            LOOP -> return wrap(position, length)
            PING_PONG -> {
                val phase: Double = wrap(position, length * 2)
                return if (phase <= length) phase else length * 2 - phase
            }

            STEP -> {
                val step = stepSeconds.toDouble()
                return floor(position / step) * step
            }

            else -> return position
        }
    }

    fun envelope(seconds: Double, rank: Int, count: Int, seed: Long): Float {
        val cycle = seconds / periodSeconds
        val phase: Double = wrap(cycle, 1.0)
        val envelope: Float
        when (style) {
            PULSE -> envelope = (.5 - .5 * cos(phase * Math.PI * 2)).toFloat()
            SWELL -> envelope = phase.toFloat()
            BURST ->
                if (
                    FaultModel.random(seed xor FaultModel.mix(floor(cycle).toLong())) >= chance ||
                        phase >= width
                )
                    envelope = 0f
                else
                    envelope =
                        FaultModel.clamp(
                            min(phase / .025, (width - phase) / .075).toFloat(),
                            0f,
                            1f,
                        )

            CASCADE -> {
                var distance = abs(phase - rank / max(1, count).toDouble())
                distance = min(distance, 1 - distance)
                envelope = max(0.0, 1 - distance * max(1, count)).toFloat()
            }

            else -> return 1f
        }
        return 1 - depth + depth * envelope
    }

    fun warped(): Boolean {
        return clock != FREE || speed != 1f || hold
    }

    companion object {
        const val NATURAL: Int = 0
        const val PULSE: Int = 1
        const val SWELL: Int = 2
        const val BURST: Int = 3
        const val CASCADE: Int = 4
        const val FREE: Int = 0
        const val LOOP: Int = 1
        const val PING_PONG: Int = 2
        const val STEP: Int = 3

        fun defaults(): LivePerformance {
            return LivePerformance(NATURAL, FREE, 1f, 90f, .75f, 4f, 4f, .3f, .65f, false)
        }

        fun wrap(value: Double, span: Double): Double {
            return value - floor(value / span) * span
        }
    }
}
